### DISCUSSION TITLE
Manual de Estilo para Formatação Automática de Código em Potigol

### DISCUSSION THREAD

**Discussion by leonelsanchesdasilva** [potigol/potigol#63] [Q&A] [answered]
Vamos escrever na nossa implementação de Potigol em TypeScript (https://github.com/DesignLiquido/potigol) um formatador automático de código. A ideia é pressionar Alt/Option + Shift + F no VSCode e o código ficar formatado bonitinho. Para isso, precisamos entender algumas coisas.

Pelo que acompanho dos posts do X do Professor Leonardo, mais o que temos em https://github.com/potigol/beecrowd, podemos depreender as seguintes diretrizes de um hipotético manual de estilo:

- Dois espaços à frente de cada novo bloco de escopo;
- Linhas com atribuições não maiores que 80 caracteres. Se a linha é maior que isso, o lado direito da atribuição fica numa nova linha, com dois espaços à frente;
- Condição em instrução `se`, se for maior que 80 caracteres, quebra de linha em conectores lógicos (https://github.com/potigol/beecrowd/blob/master/src/3100/3134.poti)
- Um espaço entre operandos e operadores, sejam eles matemáticos, lógicos, mas não unários;
- Chamadas de métodos encadeados sempre quebram linha (por exemplo, https://github.com/potigol/beecrowd/blob/master/src/3100/3159.poti)

São estes? Ou esqueci de algum?

**Comment by lrlucena** [comment, marked as answer]
Olá @leonelsanchesdasilva 

Ótima iniciativa. Eu vou colocar aqui mais algumas recomendações de estilo. 

Nas primeiras versões do Potigol havia um editor ([ePotigol](https://github.com/potigol/EditorPotigol)) com recursos de [formatação automática](https://github.com/potigol/EditorPotigol/blob/master/src/main/scala/br/edu/ifrn/potigol/editor/PrettyListener.java). O editor tinha alguns bugs e eu acabei deixando de manter, mas a formatação automática funcionava. Eu vou pegar alguns exemplos formatados para colocar aqui.

Ultimamente andei avaliando alternativas ao estilo padrão. Um exemplo é o `se`.

**Opção 1**
Comando
```portugol
se n mod 2 == 0 entao
  escreva "par"
senao
  escreva "ímpar"
fim
```

Expressão
```portugol
x =
  se n mod 2 == 0 então
    "par"
  senão
     "ímpar"
  fim
```

**Opção 2**
Comando
```portugol
se n mod 2 == 0
  entao escreva "par"
  senao escreva "ímpar"
fim
```

Expressão
```portugol
x = se n mod 2 == 0
  entao "par"
  senao "ímpar"
fim
```

**Opção 3 - Expressão**
```portugol
x = se n mod 2 == 0 entao "par" senao "ímpar" fim
```

**Reply by leonelsanchesdasilva** [reply]
Oi, professor. Obrigado pela dica do `PrettyListener`. Nós tentamos evitar o uso do ANTLR na nossa implementação por vários motivos, mas este arquivo é perfeito para orientar nossa implementação. Os exemplos também são ótimos!

Abri uma issue no nosso repositório e compartilho aqui o link para acompanhamento: https://github.com/DesignLiquido/potigol/issues/2
